client: Cache generated X509 certificate

rohansingh opened the pull request:
If we generate an X509 certificate and get it signed by ssh-agent, cache
the resulting certificate and key in ~/.helios. This can reduce the time it
takes for future requests by around a second since generating new key pairs
is an expensive operation.

rohansingh commented:
@spotify/helios-team

mattnworb commented on helios-client/src/main/java/com/spotify/helios/client/HttpsHandlers.java:
```diff
@@ -92,10 +122,85 @@ protected CertificateAndPrivateKey createCertificateAndPrivateKey() {
+        }
+      } catch (IOException | GeneralSecurityException e) {
+        // some sort of issue with cached certificate, that's fine
+        log.debug("error reading cache certificate and key for identity={}", identity.getComment());
```
logging `e.getMessage()` might be useful too

mattnworb commented:
I think the logic looks ok but it would be nice if there was a test of the cached-file-writing and -reading too, perhaps by extracting that logic out of the ssh handler and into it's own class.

davidxia commented on helios-client/src/main/java/com/spotify/helios/client/HttpsHandlers.java:
```diff
@@ -92,10 +122,85 @@ protected CertificateAndPrivateKey createCertificateAndPrivateKey() {
+
+        final String certData = X509CertificateFactory.CERT_ENCODING
+            .encode(certificateAndKeyPair.getCertificate().getEncoded());
+        final String certPem = String.format("%1$s\n%2$s\n%3$s",
```
Can't we just use ordinary indexing here, ie `%s\n%s\n%s`? I actually didn't know about explicit indexing syntax, but it seems ordinary indexing is more readable.

rohansingh replied:
Heh. I actually didn't know about that syntax.

davidxia commented on helios-client/src/main/java/com/spotify/helios/client/HttpsHandlers.java:
```diff
@@ -92,10 +122,85 @@ protected CertificateAndPrivateKey createCertificateAndPrivateKey() {
+      CertificateAndPrivateKey cached = null;
+
+      try {
+        if (Files.exists(cacheCertrificatePath) && Files.exists(cacheKeyPath)) {
```
s/cacheCertrificatePath/cacheCertificatePath/

davidxia commented on helios-client/src/main/java/com/spotify/helios/client/HttpsHandlers.java:
```diff
@@ -92,10 +122,85 @@ protected CertificateAndPrivateKey createCertificateAndPrivateKey() {
+        final String certData = X509CertificateFactory.CERT_ENCODING
+            .encode(certificateAndKeyPair.getCertificate().getEncoded());
+        final String certPem = String.format("%1$s\n%2$s\n%3$s",
+                                             "-----BEGIN CERTIFICATE-----",
```
Make this and END CERT public static final Strings in X509CertificateFactory?

davidxia commented:
Yea it'd be nice to have tests for this :)

gimaker commented:
I would like to see a test for how this behaves when the cached certificate is expired. From what I gather from the code it looks like it would just fail without re-generating the cert??

rohansingh commented:
@gimaker Nah, check out https://github.com/spotify/helios/pull/847/files#diff-65069841b102a37ed7d79b23a0b97e02R190. If the cached certificate is expired, the cache will not be used and instead it will fallback to normal mode.

But yeah I'll add tests.

rohansingh commented:
Please re-review, I've taken a completely different and (I think) cleaner approach where the caching is handled entirely inside `X509CertificateFactory`.

mattnworb commented:
Overall :+1: and this definitely looks cleaner

rohansingh commented:
@mattnworb Thanks, addressed your comments — added & fixed logging and the tests.

@spotify/helios-team Going to merge this now (once the tests pass) unless anyone has any objections.

davidxia commented:
I haven't gotten a chance to read the whole thing yet, but do we want another test for `CertificateAndPrivateKey`?

mattnworb commented:
Oh lame, because I commented on the commit itself my comments are lost, dumb github

:+1:

rohansingh commented:
@dxia You mean for `CertificateAndPrivateKey.from`? Because that's the only logic in there.

rohansingh commented:
For the record, not writing a test for `CertificateAndPrivateKey.from` since the logic of reading PEM files from disk is covered by `X509CertificateFactoryTest`.
